python/ray/data/impl/block_batching.py

```diff
@@ -0,0 +1,130 @@
+import collections
+import itertools
+from typing import Iterator, Iterable, Union, Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import pyarrow
+    import pandas
+
+import numpy as np
+
+import ray
+from ray.types import ObjectRef
+from ray.data.block import Block, BlockAccessor
+from ray.data.impl.batcher import Batcher
+from ray.data.impl.stats import DatasetStats, DatasetPipelineStats
+
+# An output type of iter_batches() determined by the batch_format parameter.
+BatchType = Union["pandas.DataFrame", "pyarrow.Table", np.ndarray, list]
+
+
+def batch_blocks(
+    blocks: Iterator[Block],
+    stats: Union[DatasetStats, DatasetPipelineStats],
```

Comment:
Note that we take either `DatasetStats` or `DatasetPipelineStats` and expect them to implement the same iterator stats interface. In a future PR, we might want to look at unifying this interface under something like a `DatasetIteratorStats` object that `DatasetStats` and `DatasetPipelineStats` each have an instance of.